spf/resolver.go

```diff
@@ -0,0 +1,81 @@
+package spf
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net"
+
+	"github.com/pkg/errors"
+)
+
+// This file includes all the DNS Resolvers used by package spf.
+
+// DnsResolver looks up txt strings associated with a FQDN.
+type DnsResolver interface {
+	GetTxt(string) []string // Given a DNS label, return the TXT values records.
+}
+
+// The "Live DNS" Resolver:
+
+type dnsLive struct {
+	filename string
+	cache    dnsCache
+}
+
+func NewResolverLive(filename string) *dnsLive {
+	// Does live DNS lookups. Records them. Writes file on Close.
+	c := &dnsLive{filename: filename}
+	return c
+}
+
+func (c *dnsLive) GetTxt(label string) ([]string, error) {
+	// Try the cache.
+	txts, ok := c.cache.dnsGet(label, "txt")
+	if ok {
+		return txts, nil
+	}
+
+	// Populate the cache:
+	t, err := net.LookupTXT(label)
+	if err == nil {
+		c.cache.dnsPut(label, "txt", t)
+	}
+
+	return t, err
+}
+
+func (c *dnsLive) Close() {
+	// Write out and close the file.
+	m, _ := json.MarshalIndent(c.cache, "", "  ")
+	m = append(m, "\n"...)
+	ioutil.WriteFile(c.filename, m, 0666)
+}
+
+// The "Pre-Cached DNS" Resolver:
+
+type dnsPreloaded struct {
```

Comment:
this type may not be neccessary. If we just implement `GetTxt` on the `dnsCache` itself, it can serve as a resolver.